statsmodels/sandbox/mice/mice.py

```diff
@@ -0,0 +1,1287 @@
+    Notes
+    -----
+    Allowed perturbation methods are 'gaussian' (the model parameters
+    are set to a draw from the Gaussian approximation to the posterior
+    distribution), and 'boot' (the model parameters are set to the
+    estimated values obtained when fitting a bootstrapped version of
+    the data set).
+
+    `history_callback` can be implemented to have side effects such as
+    saving the current imputed data set to disk.
+    """ % {'_mice_data_example_1': _mice_data_example_1,
+           '_mice_data_example_2': _mice_data_example_2}
+
+    def __init__(self, data, perturbation_method='gaussian',
+                 k_pmm=20, history_callback=None):
+
+        # Drop observations where all variables are missing.  This
+        # also has the effect of copying the data frame.
+        self.data = data.dropna(how='all').reset_index(drop=True)
+
+        self.history_callback = history_callback
+        self.history = []
+        self.predict_kwds = {}
+
+        # Assign the same perturbation method for all variables.
+        # Can be overriden when calling 'set_imputer'.
+        self.perturbation_method = defaultdict(lambda :
+                                               perturbation_method)
+
+        # Map from variable name to indices of observed/missing
+        # values.
+        self.ix_obs = {}
+        self.ix_miss = {}
+        for col in self.data.columns:
+            ix_obs, ix_miss = self._split_indices(self.data[col])
+            self.ix_obs[col] = ix_obs
+            self.ix_miss[col] = ix_miss
+
+        # Most recent model instance and results instance for each variable.
+        self.models = {}
+        self.results = {}
+
+        # Map from variable names to the conditional formula.
+        self.conditional_formula = {}
+
+        # Map from variable names to init/fit args of the conditional
+        # models.
+        self.init_kwds = defaultdict(lambda : dict())
+        self.fit_kwds = defaultdict(lambda : dict())
+
+        # Map from variable names to the model class.
+        self.model_class = {}
+
+        # Map from variable names to most recent params update.
+        self.params = {}
+
+        # Set default imputers.
+        for vname in data.columns:
+            self.set_imputer(vname)
+
+        # The order in which variables are imputed in each cycle.
+        # Impute variables with the fewest missing values first.
+        vnames =list(data.columns)
+        nmiss = [len(self.ix_miss[v]) for v in vnames]
+        nmiss = np.asarray(nmiss)
+        ii = np.argsort(nmiss)
+        ii = ii[sum(nmiss == 0):]
+        self._cycle_order = [vnames[i] for i in ii]
+
+        self._initial_imputation()
+
+        self.k_pmm = k_pmm
+
+
+    def _initial_imputation(self):
+        """
+        Use a PMM-like procedure for initial imputed values.
+        """
+
+        def f(vec):
+            vec = vec.dropna()
+            mn = vec.mean()
+            ix = np.argmin(vec - mn)
```

Comment:
shouldn't this be absolute value?
IIUC, it's the element closest to the mean, in abs norm.